File: iosApp/iosApp/Services/MigrationManager.swift
import Foundation
import Supabase

/// Résultat de l'application d'une migration
struct MigrationResult {
    let success: Bool
    let alreadyApplied: Bool
    let message: String
    let executionTimeMs: Int?
    let error: String?
}

/// Résultat de l'exécution des migrations
struct MigrationRunResult {
    let success: Bool
    let migrationsApplied: [String]
    let migrationsFailed: [(String, String)] // name, error message
    let migrationsSkipped: [String]
    let systemNotInstalled: Bool
    let errorMessage: String?
}

/// Résultat de la vérification de compatibilité app/DB
enum CompatibilityResult {
    case compatible
    case appTooOld(appVersion: Int, minRequired: Int, dbVersion: Int)
    case unknown(reason: String)
}

/// DTO pour les migrations stockées dans schema_migrations
struct SchemaMigrationDto: Codable {
    let name: String
    let checksum: String?
    let appliedAt: Int64?
    let appliedBy: String?
    let success: Bool
    let executionTimeMs: Int?
    let errorMessage: String?

    enum CodingKeys: String, CodingKey {
        case name
        case checksum
        case appliedAt = "applied_at"
        case appliedBy = "applied_by"
        case success
        case executionTimeMs = "execution_time_ms"
        case errorMessage = "error_message"
    }
}

/// DTO pour la version du schéma
struct SchemaVersionDto: Codable {
    let schemaVersion: Int
    let minAppVersion: Int
    let updatedAt: Int64?

    enum CodingKeys: String, CodingKey {
        case schemaVersion = "schema_version"
        case minAppVersion = "min_app_version"
        case updatedAt = "updated_at"
    }
}

/// Gestionnaire des migrations de schéma Supabase pour iOS
///
/// Ce gestionnaire:
/// - Charge les fichiers SQL depuis le bundle de l'app
/// - Compare avec les migrations déjà appliquées dans Supabase
/// - Applique les nouvelles migrations dans l'ordre alphabétique
class MigrationManager {

    /// Version du schéma supportée par cette version de l'app
    static let APP_SCHEMA_VERSION = 2

    private let supabase: SupabaseClient

    init() {
        self.supabase = SupabaseService.shared.client
    }

    // MARK: - Compatibility Check

    /// Vérifie si cette version de l'app est compatible avec la base de données
    func checkCompatibility() async -> CompatibilityResult {
        do {
            let response: [SchemaVersionDto] = try await supabase.database
                .rpc("get_schema_version")
                .execute()
                .value

            guard let schemaVersion = response.first else {
                print("⚠️ Système de versioning non installé - compatibilité assumée")
                return .compatible
            }

            let dbVersion = schemaVersion.schemaVersion
            let minAppVersion = schemaVersion.minAppVersion

            if Self.APP_SCHEMA_VERSION < minAppVersion {
                print("❌ App trop ancienne: app=\(Self.APP_SCHEMA_VERSION), min=\(minAppVersion), db=\(dbVersion)")
                return .appTooOld(appVersion: Self.APP_SCHEMA_VERSION, minRequired: minAppVersion, dbVersion: dbVersion)
            } else {
                print("✅ App compatible: app=\(Self.APP_SCHEMA_VERSION), min=\(minAppVersion), db=\(dbVersion)")
                return .compatible
            }
        } catch {
            print("⚠️ Impossible de vérifier la compatibilité: \(error.localizedDescription)")
            return .unknown(reason: error.localizedDescription)
        }
    }

    // MARK: - Load Migrations

    /// Charge toutes les migrations depuis le bundle de l'app
    func loadMigrationsFromBundle() -> [(name: String, sql: String, checksum: String)] {
        var migrations: [(name: String, sql: String, checksum: String)] = []

        guard let bundlePath = Bundle.main.resourcePath else {
            print("❌ Impossible de trouver le chemin du bundle")
            return []
        }

        let migrationsPath = (bundlePath as NSString).appendingPathComponent("migrations")
        let fileManager = FileManager.default

        do {
            let files = try fileManager.contentsOfDirectory(atPath: migrationsPath)
                .filter { $0.hasSuffix(".sql") }
                .sorted()

            for fileName in files {
                let filePath = (migrationsPath as NSString).appendingPathComponent(fileName)
                if let sql = try? String(contentsOfFile: filePath, encoding: .utf8) {
                    let name = String(fileName.dropLast(4)) // Remove .sql
                    let checksum = sql.md5()
                    migrations.append((name: name, sql: sql, checksum: checksum))
                }
            }
        } catch {
            print("❌ Erreur lors du listage des migrations: \(error.localizedDescription)")
        }

        return migrations
    }

    // MARK: - Get Applied Migrations

    /// Récupère les migrations déjà appliquées depuis Supabase
    func getAppliedMigrations() async -> [SchemaMigrationDto] {
        do {
            let response: [SchemaMigrationDto] = try await supabase.database
                .rpc("get_applied_migrations")
                .execute()
                .value
            return response
        } catch {
            print("❌ Erreur lors de la récupération des migrations: \(error.localizedDescription)")
            return []
        }
    }

    /// Vérifie si le système de migration est installé
    func isMigrationSystemInstalled() async -> Bool {
        do {
            let _: [SchemaMigrationDto] = try await supabase.database
                .rpc("get_applied_migrations")
                .execute()
                .value
            return true
        } catch {
            print("⚠️ Système de migration non installé: \(error.localizedDescription)")
            return false
        }
    }

    // MARK: - Apply Migration

    /// Applique une migration
    func applyMigration(name: String, sql: String, checksum: String?, appliedBy: String = "ios_app") async -> MigrationResult {
        do {
            let params: [String: AnyJSON] = [
                "p_name": .string(name),
                "p_sql": .string(sql),
                "p_checksum": checksum.map { .string($0) } ?? .null,
                "p_applied_by": .string(appliedBy)
            ]

            let response = try await supabase.database
                .rpc("apply_migration", params: params)
                .execute()

            // Parse JSON response
            if let data = response.data,
               let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any] {
                return MigrationResult(
                    success: json["success"] as? Bool ?? false,
                    alreadyApplied: json["already_applied"] as? Bool ?? false,
                    message: json["message"] as? String ?? "Unknown result",
                    executionTimeMs: json["execution_time_ms"] as? Int,
                    error: json["error"] as? String
                )
            }

            return MigrationResult(success: false, alreadyApplied: false, message: "Failed to parse response", executionTimeMs: nil, error: nil)
        } catch {
            print("❌ Erreur lors de l'application de la migration \(name): \(error.localizedDescription)")
            return MigrationResult(
                success: false,
                alreadyApplied: false,
                message: "Failed to apply migration: \(error.localizedDescription)",
                executionTimeMs: nil,
                error: error.localizedDescription
            )
        }
    }

    // MARK: - Run Pending Migrations

    /// Récupère les migrations en attente
    func getPendingMigrations() async -> [(name: String, sql: String, checksum: String)] {
        let allMigrations = loadMigrationsFromBundle()
        let appliedMigrations = await getAppliedMigrations()
            .filter { $0.success }
            .map { $0.name }

        return allMigrations.filter { !appliedMigrations.contains($0.name) }
    }

    /// Exécute toutes les migrations en attente
    func runPendingMigrations(appliedBy: String = "ios_app") async -> MigrationRunResult {
        // Vérifier si le système est installé
        let systemInstalled = await isMigrationSystemInstalled()
        if !systemInstalled {
            return MigrationRunResult(
                success: false,
                migrationsApplied: [],
                migrationsFailed: [],
                migrationsSkipped: [],
                systemNotInstalled: true,
                errorMessage: "Migration system not installed in Supabase. Please run 2026011701_migration_system.sql first."
            )
        }

        let pendingMigrations = await getPendingMigrations()

        if pendingMigrations.isEmpty {
            return MigrationRunResult(
                success: true,
                migrationsApplied: [],
                migrationsFailed: [],
                migrationsSkipped: [],
                systemNotInstalled: false,
                errorMessage: nil
            )
        }

        print("📦 \(pendingMigrations.count) migration(s) en attente")

        var applied: [String] = []
        var failed: [(String, String)] = []
        var skipped: [String] = []

        for migration in pendingMigrations {
            print("⏳ Application de \(migration.name)...")

            let result = await applyMigration(
                name: migration.name,
                sql: migration.sql,
                checksum: migration.checksum,
                appliedBy: appliedBy
            )

            if result.alreadyApplied {
                print("⏭️ \(migration.name) déjà appliquée")
                skipped.append(migration.name)
            } else if result.success {
                print("✅ \(migration.name) appliquée en \(result.executionTimeMs ?? 0)ms")
                applied.append(migration.name)
            } else {
                print("❌ \(migration.name) échouée: \(result.error ?? "Unknown error")")
                failed.append((migration.name, result.error ?? "Unknown error"))

                // Arrêter en cas d'échec
                return MigrationRunResult(
                    success: false,
                    migrationsApplied: applied,
                    migrationsFailed: failed,
                    migrationsSkipped: skipped,
                    systemNotInstalled: false,
                    errorMessage: "Migration \(migration.name) failed: \(result.error ?? "Unknown error")"
                )
            }
        }

        return MigrationRunResult(
            success: true,
            migrationsApplied: applied,
            migrationsFailed: failed,
            migrationsSkipped: skipped,
            systemNotInstalled: false,
            errorMessage: nil
        )
    }
}

// MARK: - String MD5 Extension

extension String {
    func md5() -> String {
        let data = Data(self.utf8)
        var digest = [UInt8](repeating: 0, count: Int(CC_MD5_DIGEST_LENGTH))
        _ = data.withUnsafeBytes {
            CC_MD5($0.baseAddress, CC_LONG(data.count), &digest)
        }
        return digest.map { String(format: "%02x", $0) }.joined()
    }
}

// Import CommonCrypto for MD5
import CommonCrypto

File: app/src/main/java/com/medistock/data/migration/MigrationManager.kt
package com.medistock.data.migration

import android.content.Context
import com.medistock.data.remote.repository.MigrationRepository
import com.medistock.data.remote.repository.MigrationResult
import com.medistock.shared.domain.compatibility.CompatibilityChecker
import com.medistock.shared.domain.compatibility.CompatibilityResult
import com.medistock.shared.domain.compatibility.SchemaVersion
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import java.security.MessageDigest

/**
 * Représente une migration SQL à appliquer
 */
data class Migration(
    val name: String,
    val sql: String,
    val checksum: String
) {
    companion object {
        /**
         * Extrait le nom de la migration depuis le nom du fichier
         * Ex: "2026011702_add_feature.sql" -> "2026011702_add_feature"
         */
        fun nameFromFileName(fileName: String): String {
            return fileName.removeSuffix(".sql")
        }
    }
}

/**
 * Résultat de l'exécution des migrations
 */
data class MigrationRunResult(
    val success: Boolean,
    val migrationsApplied: List<String>,
    val migrationsFailed: List<Pair<String, String>>, // name to error message
    val migrationsSkipped: List<String>, // already applied
    val systemNotInstalled: Boolean = false,
    val errorMessage: String? = null
)

/**
 * Callback pour suivre la progression des migrations
 */
interface MigrationProgressListener {
    fun onMigrationStart(name: String, current: Int, total: Int)
    fun onMigrationComplete(name: String, result: MigrationResult)
    fun onAllComplete(result: MigrationRunResult)
}

/**
 * Gestionnaire des migrations de schéma Supabase
 *
 * Ce gestionnaire:
 * - Charge les fichiers SQL depuis assets/migrations/
 * - Compare avec les migrations déjà appliquées dans Supabase
 * - Applique les nouvelles migrations dans l'ordre alphabétique
 *
 * Usage:
 * ```kotlin
 * val manager = MigrationManager(context)
 * val result = manager.runPendingMigrations()
 * if (result.success) {
 *     println("${result.migrationsApplied.size} migrations applied")
 * }
 * ```
 */
class MigrationManager(
    private val context: Context,
    private val repository: MigrationRepository = MigrationRepository()
) {
    companion object {
        private const val TAG = "MigrationManager"
        private const val MIGRATIONS_FOLDER = "migrations"

        /**
         * Version du schéma supportée par cette version de l'app.
         * @see CompatibilityChecker.APP_SCHEMA_VERSION for the actual value
         */
        val APP_SCHEMA_VERSION: Int
            get() = CompatibilityChecker.APP_SCHEMA_VERSION
    }

    /**
     * Vérifie si cette version de l'app est compatible avec la base de données.
     *
     * Cette vérification doit être faite AVANT de tenter d'utiliser l'app.
     * Si le résultat est AppTooOld, l'utilisateur doit mettre à jour l'app.
     *
     * @return CompatibilityResult indiquant si l'app peut être utilisée
     */
    suspend fun checkCompatibility(): CompatibilityResult {
        return try {
            val schemaVersionDto = repository.getSchemaVersion()

            // Convert DTO to shared model
            val schemaVersion = schemaVersionDto?.let {
                SchemaVersion(
                    schemaVersion = it.schemaVersion,
                    minAppVersion = it.minAppVersion,
                    updatedAt = it.updatedAt
                )
            }

            // Use shared compatibility checker
            val result = CompatibilityChecker.checkCompatibility(schemaVersion)

            // Log the result
            println(CompatibilityChecker.formatCompatibilityInfo(result))

            result
        } catch (e: Exception) {
            println("⚠️ Impossible de vérifier la compatibilité: ${e.message}")
            CompatibilityResult.Unknown(e.message ?: "Unknown error")
        }
    }

    /**
     * Charge toutes les migrations depuis assets/migrations/
     * Les fichiers doivent suivre le format: YYYYMMDDNN_description.sql
     */
    suspend fun loadMigrationsFromAssets(): List<Migration> = withContext(Dispatchers.IO) {
        try {
            val assetManager = context.assets
            val files = assetManager.list(MIGRATIONS_FOLDER) ?: emptyArray()

            files
                .filter { it.endsWith(".sql") }
                .sorted() // Tri alphabétique = ordre chronologique avec le format YYYYMMDDNN
                .mapNotNull { fileName ->
                    try {
                        val sql = assetManager.open("$MIGRATIONS_FOLDER/$fileName")
                            .bufferedReader()
                            .use { it.readText() }

                        Migration(
                            name = Migration.nameFromFileName(fileName),
                            sql = sql,
                            checksum = calculateMD5(sql)
                        )
                    } catch (e: Exception) {
                        println("⚠️ Erreur lors du chargement de $fileName: ${e.message}")
                        null
                    }
                }
        } catch (e: Exception) {
            println("❌ Erreur lors du listage des migrations: ${e.message}")
            emptyList()
        }
    }

    /**
     * Récupère les migrations qui n'ont pas encore été appliquées
     */
    suspend fun getPendingMigrations(): List<Migration> {
        val allMigrations = loadMigrationsFromAssets()
        val appliedMigrations = repository.getAppliedMigrations()
            .filter { it.success }
            .map { it.name }
            .toSet()

        return allMigrations.filter { it.name !in appliedMigrations }
    }

    /**
     * Exécute toutes les migrations en attente
     *
     * IMPORTANT: This method now requires authentication. Pass the Supabase Auth
     * access token to use the secure Edge Function for migration execution.
     * If no token is provided, falls back to direct RPC (legacy mode).
     *
     * @param listener Callback optionnel pour suivre la progression
     * @param appliedBy Identifiant de l'utilisateur/app qui applique les migrations
     * @param accessToken Supabase Auth access token for Edge Function authentication
     */
    suspend fun runPendingMigrations(
        listener: MigrationProgressListener? = null,
        appliedBy: String = "app",
        accessToken: String? = null
    ): MigrationRunResult {
        // Vérifier si le système de migration est installé
        if (!repository.isMigrationSystemInstalled()) {
            val result = MigrationRunResult(
                success = false,
                migrationsApplied = emptyList(),
                migrationsFailed = emptyList(),
                migrationsSkipped = emptyList(),
                systemNotInstalled = true,
                errorMessage = "Migration system not installed in Supabase. Please run 2026011701_migration_system.sql first."
            )
            listener?.onAllComplete(result)
            return result
        }

        val pendingMigrations = getPendingMigrations()

        if (pendingMigrations.isEmpty()) {
            val result = MigrationRunResult(
                success = true,
                migrationsApplied = emptyList(),
                migrationsFailed = emptyList(),
                migrationsSkipped = emptyList()
            )
            listener?.onAllComplete(result)
            return result
        }

        println("Pending migrations: ${pendingMigrations.size} migration(s)")
        if (accessToken != null) {
            println("Using Edge Function for secure migration execution")
        } else {
            println("Warning: No access token provided, using legacy RPC method")
        }

        val applied = mutableListOf<String>()
        val failed = mutableListOf<Pair<String, String>>()
        val skipped = mutableListOf<String>()

        pendingMigrations.forEachIndexed { index, migration ->
            listener?.onMigrationStart(migration.name, index + 1, pendingMigrations.size)
            println("Applying migration: ${migration.name}...")

            val result = repository.applyMigration(
                name = migration.name,
                sql = migration.sql,
                checksum = migration.checksum,
                appliedBy = appliedBy,
                accessToken = accessToken
            )

            listener?.onMigrationComplete(migration.name, result)

            when {
                result.alreadyApplied -> {
                    println("Skipped: ${migration.name} (already applied)")
                    skipped.add(migration.name)
                }
                result.success -> {
                    println("Success: ${migration.name} applied in ${result.executionTimeMs}ms")
                    applied.add(migration.name)
                }
                else -> {
                    println("Failed: ${migration.name} - ${result.error}")
                    failed.add(migration.name to (result.error ?: "Unknown error"))
                    // On arrête en cas d'échec pour éviter les problèmes de dépendances
                    val runResult = MigrationRunResult(
                        success = false,
                        migrationsApplied = applied,
                        migrationsFailed = failed,
                        migrationsSkipped = skipped,
                        errorMessage = "Migration ${migration.name} failed: ${result.error}"
                    )
                    listener?.onAllComplete(runResult)
                    return runResult
                }
            }
        }

        val runResult = MigrationRunResult(
            success = true,
            migrationsApplied = applied,
            migrationsFailed = failed,
            migrationsSkipped = skipped
        )
        listener?.onAllComplete(runResult)
        return runResult
    }

    /**
     * Vérifie l'état des migrations sans les appliquer
     *
     * @return Triple de (appliquées, en attente, système installé)
     */
    suspend fun checkMigrationStatus(): Triple<List<String>, List<String>, Boolean> {
        val systemInstalled = repository.isMigrationSystemInstalled()

        if (!systemInstalled) {
            return Triple(emptyList(), emptyList(), false)
        }

        val allMigrations = loadMigrationsFromAssets().map { it.name }
        val appliedMigrations = repository.getAppliedMigrations()
            .filter { it.success }
            .map { it.name }

        val pendingMigrations = allMigrations.filter { it !in appliedMigrations }

        return Triple(appliedMigrations, pendingMigrations, true)
    }

    /**
     * Calcule le hash MD5 d'une chaîne
     */
    private fun calculateMD5(input: String): String {
        val md = MessageDigest.getInstance("MD5")
        val digest = md.digest(input.toByteArray())
        return digest.joinToString("") { "%02x".format(it) }
    }
}
